### PR TITLE
[Luxon] Provide generics to allow individual types to be valid/invalid

### PR DESCRIPTION
**Pull request by CarsonF**
Following up on #64995

In certain situations it's possible to know statically that a `DateTime` (etc) is valid, even without the `throwOnInvalid` enabled.

For example, it's probably common to do
```ts
DateTime.now().toISO()
```
This will always return a string, as now() always returns a valid DateTime.
This PR enables TS to know that.

### CODE REVIEW

**Comment by jameyg42** [comment]
does `Interval` have similar characteristics as `now()`?   It almost looks like the only way to create an invalid Interval (which introduces nulls) is with `fromISO()` which is why i never realized things like `Interval.start|end` were nullable.

**Comment by CarsonF** [comment]
Possibly a solution to the `isValid` check narrowing and the generic being introduced:
```ts
const dt = DateTime.fromISO('') as any as DateTime<true> | DateTime<false>;
dt.toISO() // string | null
if (dt.isValid) {
    dt.toISO() // string
}
```
Since `DateTime<true>.isValid: true` and `DateTime<false>.isValid: false`, TS can narrow that type union based on that discriminator.

Idk if it would make sense to change the names to something like:
```ts
class DateTimeClass<...> {}

type DateTime = DateTimeClass<true> | DateTimeClass<false>;
```
That probably wouldn't work. 

Maybe something like
```ts
type DateTimeValidityUnion = ValidDateTime | InvalidDateTime;

type ValidDateTime = DateTime<true>;
type InvalidDateTime = DateTime<false>;
```
Then all the factory methods would return 1 of those 3 types.

**Review by koooge — COMMENTED** [review]
As for cron, only the comments have been changed. It is not clear if it is necessary.

**Comment by danqing** [comment]
It would be great to see some momentum on this PR. @CarsonF is there more work needed here?

**Comment by CarsonF** [comment]
@danqing I'm just not confident of the implications of this change. And after the backlash from #64995, I don't want to add another breaking change pain point.

**Comment by pvogel1967** [comment]
@CarsonF elegant approach!  I'd be willing to give this a spin in our code that suffered badly from #64995 .  Do you have sense of where you'd be concerned about the implications of the change?

**Comment by pvogel1967** [comment]
Honestly, I love the _idea_ behind the #64995 change as it highlights a lurking issue via type checking, always a good thing.  But with that _must_ come a clean way to assert validity per standard type guards, etc. or you wind up peppering your code with `!` which anesthetizes people to what is, in general, a very bad practice.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/luxon/src/_util.d.ts
1. types/luxon/test/luxon-tests.module.ts
1. types/luxon/src/datetime.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9fb9e16b2458774848abefd0028585e86d5d06f3
</p>

**Comment by CarsonF** [comment]
Just got the branch rebased with all the formatting changes that were conflicting.
I think I have an idea for narrowing with `isValid`.

@pvogel1967 I think mainly the concern is just around just being able to drop this change in and it be the same. i.e. I haven't added required generics that need to be added everywhere. So type usage doesn't need to change at all in your repos.

**Comment by danqing** [comment]
Yea I agree with @pvogel1967 and our codebase has also suffered from #64995 :)

**Comment by CarsonF** [comment]
Curious, why don't you guys want to enable `throwOnInvalid`?

**Comment by pvogel1967** [comment]
I prefer, when using DateTime to _validate_ data from the user to _not_ follow an exception-handling path, but to just check if it's valid, and if not, to use invalidExplanation and invalidReason in an error message.

**Comment by CarsonF** [comment]
If you want to copy this down to your repo @pvogel1967, I'd be curious to hear your feedback. I think the branch is in a much better spot now. I've handled the type narrowing on `isValid` too.

I think there's only some minor spots now where I haven't handled the interaction between the existing validity and the validity of method args and how that affects the output of an instance method.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/luxon/src/zone.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d40ceaa9403e483bcdc989b41ad36eb1f0a3c902
</p>

**Comment by andrewbranch** [comment]
Is this still waiting on additional feedback from @pvogel1967?

**Comment by CarsonF** [comment]
> Is this still waiting on additional feedback from @pvogel1967?

I think it would be good to get feedback from @pvogel1967 and/or @peterblazejewicz

**Comment by peterblazejewicz** [comment]
> > Is this still waiting on additional feedback from @pvogel1967?
> 
> I think it would be good to get feedback from @pvogel1967 and/or @peterblazejewicz

@CarsonF I'd say at your discretion, as you're actual implementation curator for the types in my opinion. The generics can be tricky (I avoid generics in our types at work, whenever we can), as the outcome of the rewrite can lead to less readable code for users, at least without at first iteration:
```ts
...: Duration<false>;
```
(what does it mean at first glance), but I understand the intent behind the change.
If we receive not positive feedback after the change, I believe the change can be iterated over quickly to adapt to users feedback.
Let's wait a day, let someone post feedback, otherwise we merge it.

**Comment by CarsonF** [comment]
@peterblazejewicz `Duration<false>` is an invalid duration. But I think the point you are making is that's not obvious, right?
We could add a type alias for readability:
```ts
type DurationValid = Duration<true>;
type DurationInvalid = Duration<false>;
```

Certainly the goal with this one is not to be breaking. I think I've achieved that, and just wanted a second pair eyes to review and maybe test.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/luxon/src/interval.d.ts
1. types/luxon/src/duration.d.ts
1. types/luxon/src/datetime.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 6272ca4cc2247d9aa1bb1e56948865ebd38f5392
</p>

**Comment by CarsonF** [comment]
I went with
```ts
type Valid = true;
type Invalid = false;

Duration<Valid>
Duration<Invalid>
```
So we don't have two more type aliases per class.

**Review by peterblazejewicz — APPROVED** [review]
@CarsonF thanks!
LGTM!

**Comment by ab-pm** [comment]
@CarsonF @pvogel1967 

> Curious, why don't you guys want to enable `throwOnInvalid`?

> I prefer, when using DateTime to *validate* data from the user to *not* follow an exception-handling path, but to just check if it's valid, and if not, to use invalidExplanation and invalidReason in an error message.

We found (after being surprised by the typing issue - which I definitely agree [is a positive thing](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65078#issuecomment-1815213252)) actually *both* usage patterns in our code base. In some parts, we used `DateTime` to validate user input, in other parts we used `DateTime` to parse API responses (and would have preferred/expected an exception). Having a global flag setting is totally annoying, and the poorest design decision on this topic.

Please share your insights on https://github.com/moment/luxon/issues/1421 to discuss how a (type)safe `moment` API could look.

File: types/luxon/src/datetime.d.ts
Comment (ab-pm) [inline types/luxon/src/datetime.d.ts:1461]: I think this should have been
```suggestion
    until(otherDateTime: DateTime): IfValid<Interval<Valid>, Interval<Invalid>, IsValid>;
```

Comment (CarsonF) [reply types/luxon/src/datetime.d.ts:1461]: https://github.com/moment/luxon/blob/master/src%2Fdatetime.js#L2081

Probably an oversight in JS src, but I felt like I should be true to it.

Comment (ab-pm) [reply types/luxon/src/datetime.d.ts:1461]: Oh wow. Reported as https://github.com/moment/luxon/issues/1551!

**Comment by philipmagnusson** [comment]
This is a breaking change in a minor version which causes some issues for us. Do you have any docs on how to migrate from version 3.3.5 to 3.3.6?

**Comment by CarsonF** [comment]
@philipmagnusson how is it breaking? I left a couple more comments [here](https://github.com/moment/luxon/issues/1419#issuecomment-1842037723), but generally I wouldn't expect any breaking changes here.

**Comment by philipmagnusson** [comment]
You've changed the type of DateTime and the return types of most methods, such as fromISO (see attached images but I guess you know about these changes). To me it's likely to be a breaking change, at least it's risky to do in a minor updates. 

Here are some of the outputs that we get from the typescript compile command after we updated from 3.3.5 to 3.3.6:

"src/components/breakMonitoring/__tests__/BreakMonitor.test.ts(48,3): error TS2322: Type 'Ref<{ readonly start: { get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; }; ... 30 more ...; mapEndpoints: (mapFn:...' is not assignable to type 
3_3_5_fromISO
3_3_6_DateTIme
3_3_6_fromISO
'DeepPartial<ComputedRef<Interval<boolean>>>'.
  Types of property 'value' are incompatible.
    Type '{ readonly start: { get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; }; ... 30 more ...; mapEndpoints: (mapFn: (d:...' is not assignable to type 'DeepPartial<Interval<boolean>>'.
      Type '{ readonly start: unknown; readonly end: unknown; readonly isValid: false; readonly invalidReason: unknown; readonly invalidExplanation: unknown; length: (unit?: keyof DurationLikeObject) => number; ... 25 more ...; mapEndpoints: (mapFn: (d: DateTime<...>) => DateTime<...>) => Interval<...> | Interval<...>; }' is not assignable to type 'DeepPartial<Interval<boolean>>'.
        Types of property 'invalidReason' are incompatible.
          Type 'unknown' is not assignable to type 'string'.
src/components/breakMonitoring/__tests__/BreakMonitor.test.ts(55,3): error TS2322: Type 'Ref<{ get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; } | { ...; }>' is not assignable to type 'DeepPartial<ComputedRef<DateTime<boolean>>>'.
  Types of property 'value' are incompatible.
    Type '{ get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; } | { ...; }' is not assignable to type 'DeepPartial<DateTime<boolean>>'.
      Type '{ get: (unit: keyof DateTime<boolean>) => number; readonly isValid: false; readonly invalidReason: unknown; readonly invalidExplanation: unknown; readonly locale: unknown; readonly numberingSystem: unknown; ... 66 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => null; }' is not assignable to type 'DeepPartial<DateTime<boolean>>'.
        Types of property 'invalidReason' are incompatible.
          Type 'unknown' is not assignable to type 'string'.
src/composables/__tests__/useBreakMonitoring.test.ts(46,2): error TS2322: Type 'Ref<{ readonly start: { get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; }; ... 30 more ...; mapEndpoints: (mapFn:...' is not assignable to type 'Ref<Interval<boolean>>'.
  Type '{ readonly start: { get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; }; ... 30 more ...; mapEndpoints: (mapFn: (d:...' is not assignable to type 'Interval<boolean>'.
    Type '{ readonly start: unknown; readonly end: unknown; readonly isValid: false; readonly invalidReason: unknown; readonly invalidExplanation: unknown; length: (unit?: keyof DurationLikeObject) => number; ... 25 more ...; mapEndpoints: (mapFn: (d: DateTime<...>) => DateTime<...>) => Interval<...> | Interval<...>; }' is not assignable to type 'Interval<boolean>'.
      Types of property 'start' are incompatible.
        Type '{}' is missing the following properties from type 'DateTime<true>': get, isValid, invalidReason, invalidExplanation, and 66 more.
src/composables/__tests__/useBreakMonitoring.test.ts(94,13): error TS2345: Argument of type 'Ref<{ readonly start: { get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; }; ... 30 more ...; mapEndpoints: (mapFn:...' is not assignable to parameter of type 'Ref<Interval<boolean>>'.
src/composables/__tests__/useBreakMonitoring.test.ts(99,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(100,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(113,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(114,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(135,13): error TS2345: Argument of type 'Ref<{ readonly start: { get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; }; ... 30 more ...; mapEndpoints: (mapFn:...' is not assignable to parameter of type 'Ref<Interval<boolean>>'.
src/composables/__tests__/useBreakMonitoring.test.ts(144,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(145,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(160,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(161,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(182,13): error TS2345: Argument of type 'Ref<{ readonly start: { get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; }; ... 30 more ...; mapEndpoints: (mapFn:...' is not assignable to parameter of type 'Ref<Interval<boolean>>'.
src/composables/__tests__/useBreakMonitoring.test.ts(187,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(188,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(219,13): error TS2345: Argument of type 'Ref<{ readonly start: { get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; }; ... 30 more ...; mapEndpoints: (mapFn:...' is not assignable to parameter of type 'Ref<Interval<boolean>>'.
src/composables/__tests__/useBreakMonitoring.test.ts(224,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(225,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(234,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(235,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(249,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(250,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(278,13): error TS2345: Argument of type 'Ref<{ readonly start: { get: (unit: keyof DateTime<boolean>) => number; readonly isValid: true; readonly invalidReason: string; readonly invalidExplanation: string; readonly locale: string; ... 67 more ...; toRelativeCalendar: (options?: ToRelativeCalendarOptions) => string; }; ... 30 more ...; mapEndpoints: (mapFn:...' is not assignable to parameter of type 'Ref<Interval<boolean>>'.
src/composables/__tests__/useBreakMonitoring.test.ts(283,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(284,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(293,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(294,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(311,36): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/composables/__tests__/useBreakMonitoring.test.ts(312,32): error TS2339: Property 'toISO' does not exist on type 'unknown'.
src/pages/provider/campaign/orderline/__tests__/Orderline.test.ts(246,3): error TS2322: Type 'DateTime<true> | DateTime<false>' is not assignable to type 'DateTime<true> | Promise<DateTime<true>>'.
  Type 'DateTime<false>' is not assignable to type 'DateTime<true> | Promise<DateTime<true>>'.
    Type 'DateTime<false>' is not assignable to type 'DateTime<true>'.
      Type 'false' is not assignable to type 'true'.
src/utils/__tests__/performanceUtils.test.ts(508,5): error TS2322: Type 'DateTime<true> | DateTime<false>' is not assignable to type 'DateTime<true> | Promise<DateTime<true>>'.
src/utils/__tests__/performanceUtils.test.ts(519,5): error TS2322: Type 'DateTime<true> | DateTime<false>' is not assignable to type 'DateTime<true> | Promise<DateTime<true>>'."

**Comment by CarsonF** [comment]
@philipmagnusson we don't have a lot of control on versioning with breaking changes. We follow the src packages versioning, so really we can only do patch releases ourselves.

It's a bit hard to know what the problem you are having is without being able to get my hands on the code.
I would expect
```ts
const breakInterval: Ref<Interval> = ref(Interval.fromISO())
```
to work fine.
Unless ref() is doing something with `DeepPartial`. I would guess that that's one incompatibility point. `DeepPartial` doesn't know which types it _shouldn't_ start modifying. If it were smarter it would know to treat `Interval` as scalar-like, and leave it untouched, but it's not. It's a bit like hacky reflection in that way.

I still recommend enabling luxon to throw on invalid input. Which does away with all of this validity generics.
```ts
import {Settings} from 'luxon';

Settings.throwOnInvalid = true;

declare module 'luxon' {
  interface TSSettings {
    throwOnInvalid: true;
  }
}
```

**Comment by philipmagnusson** [comment]
Unfortunately, `const breakInterval: Ref<Interval> = ref(Interval.fromISO())` does not work. I've had to do these type casts through our code base, for example here: 
Screenshot 2024-01-13 at 13 19 58

where `windowInterval` is a [ComputedRef](https://vuejs.org/api/reactivity-core.html#computed) https://vuejs.org/api/reactivity-core.html#computed

Screenshot 2024-01-13 at 13 20 38

I'm not sure what it would been for our code base to opt-in to `throwOnInvalid` so I'd rather not do it, but I'll look it up.

**Comment by CarsonF** [comment]
@philipmagnusson I'm assuming you are assuming all temporal inputs are valid, and not checking the `valid` prop, or `invalidReason`. The opt in just converts those "lazy invalid objects" into thrown errors. I've never actually had an error produced, since these inputs are rarely hand crafted and more rare they are invalid. So I'd strongly recommend that opt-in. It's a poor default decision from Luxon. 

Vue's `ref` and `computed` seem simple and shouldn't have any conflicts.
If you could get me a reproduction repo that would be helpful.

**Comment by philipmagnusson** [comment]
hmm... I don't understand. Should the settings you posted above remove the DateTime<true|false> definitions? I tried it but I don't think it works, or I don't understand how to insert the settings you posted aboe: 
`import {Settings} from 'luxon';

Settings.throwOnInvalid = true;

declare module 'luxon' {
  interface TSSettings {
    throwOnInvalid: true;
  }
}`

Tried to put it in ./luxon.ts file and include it in tsconfig.json and import it in the index.ts of our app as suggested here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64995#issuecomment-1580729134 and also tried to do this:

create a luxon.d.ts file: 

`export {}
declare module 'luxon' {
  interface TSSettings {
    throwOnInvalid: true
  }
}`

and do the setting:

`import { Settings } from 'luxon';

Settings.throwOnInvalid = true;

`

in the index.ts file of our app. 

The DateTime is still defined as: 

`class DateTime<IsValid extends boolean = true>`

and e.g. setLocale returns `DateTime<true>` which causes the issues for us .

**Comment by trevorr** [comment]
+1 that this is breaking change:

```
src/util/recurrence.test.ts:15:64 - error TS2345: Argument of type 'DateTime<true> | DateTime<false>' is not assignable to parameter of type 'DateTime<true> | undefined'.
  Type 'DateTime<false>' is not assignable to type 'DateTime<true>'.
    Type 'false' is not assignable to type 'true'.
```

**Comment by CarsonF** [comment]
@philipmagnusson sorry I missed your reply. The signature you posted
```ts
class DateTime<IsValid extends boolean = true>
```
Means that the `throwOnInvalid` is being applied correctly. The default validity is always `true` instead of `true` or `false` (aka `boolean`).

How is "`setLocale` returning `DateTime<true>`" an issue for you?

Would you open a new discussion or issue to continue this?

---

@trevorr "+1" comments are a frowned upon practice. It causes notification spam to maintainers and everyone participating in the thread without adding any value. Please don't do that anywhere. This is exactly what reactions are for.

Both comments & reactions are easily ignored on merged PRs.
So, please also open a new discussion or issue to continue this.

That error doesn't tell me anything without seeing the source code that produced it.
Also are you using `throwOnInvalid` or not?

**Comment by trevorr** [comment]
> without adding any value. Please don't do that anywhere

@CarsonF That's a bit condescending. It wasn't just a +1, since I included an additional error mode not yet mentioned. The error itself makes clear that the new type argument and its new usage within the interface are inherently breaking.

Here's the test that now fails to type check:

```ts
const startDate = DateTime.fromISO("2020-02-29T08:00:00Z", { zone: "utc" });
const pt2h = Duration.fromObject({ hours: 2 });

describe("formatRecurrenceRule", () => {
  it("formats none", () => {
    // error TS2345: Argument of type 'DateTime<true> | DateTime<false>' is not assignable to parameter of type 'DateTime<true> | undefined'.
    expect(formatRecurrenceRule({ type: RecurrenceType.None }, startDate, pt2h)).toBeNull();
  });
```

Here's the signature of that function:

```ts
export function formatRecurrenceRule(
  recur: Recurrence,
  startDate = DateTime.utc(),
  duration?: Duration
): string | null {
```

**Comment by CarsonF** [comment]
> The error itself makes clear that the new type argument and its new usage within the interface are inherently breaking.

I understand typescript's error message. I guess what I meant was I still needed more info about the usage.
Your new signature definitely helps.

So here's the rub.
```ts
DateTime.utc()          // always valid
DateTime.fromISO('...') // maybe valid depending on string
```
Now your `startDate` argument is inferring the type from this default value.
So TS infers that the type is _always_ a valid `DateTime`.
You can give an explicit type to tell TS that this is actually a looser type.
```ts
export function formatRecurrenceRule(
  recur: Recurrence,
  startDate: DateTime = DateTime.utc(),
  duration?: Duration
): string | null {
```
This is the same principle as
```ts
// Loosen name to any string, not just "Carson"
const name: string = "Carson";
```

This is an unfortunate consequence of this PR, but I don't see any way around it.
We are constrained by the decision of the JS library.

**Comment by trevorr** [comment]
Yeah, that totally makes sense and thanks for the explanation. That change does fix all the compile errors in my tests. Perhaps this is mainly an issue when using type inference.

Maybe what's needed is some sort of migration guide? It took a bit of digging to find this PR and figure out what was going on. Has anyone asked if the Luxon maintainers would accept some Typescript documentation and possibly even the typings into their repo? It seems like you (@CarsonF) did that in [this PR](https://github.com/moment/luxon/pull/1415) for how to configure the default behavior. Sadly, you only got some not helpful feedback from a non-maintainer.

**Comment by CarsonF** [comment]
I don't know what Issac's views on TS are. I've never seen him anywhere near these type PRs/issues. So yeah at the end of the day this is not a TS library, and TS is a second class citizen for it. So that excludes docs, breaking changes, etc related to types.
Unfortunate for sure.

**Pull request merged**
